ComSemApp/teacher/views.py

```diff
@@ -70,7 +71,59 @@ def get_context_data(self, **kwargs):
 class CourseDetailView(TeacherCourseViewMixin, DetailView):
     context_object_name = 'course'
     template_name = "ComSemApp/teacher/course.html"
+    
+    @register.filter('get_item')
+    def get_item(dictionary, key):
+        return dictionary.get(key)
 
+    def get_context_data(self, **kwargs):
+        data = super().get_context_data(**kwargs)
+
+        worksheets = Worksheet.objects.filter(course=self.course)
+        worksheetsdict = {}
+        ungradedcountdict = {}
+        attemptsdict = {}
+        for student in self.course.students.all(): 
+            worksheetcount = 0
+            ungradedcount = 0
+            attemptcount = 0
+            submissions = StudentSubmission.objects.filter(student=student)
+            for submission in submissions :
+                if submission.worksheet.course == self.course:
+                    if submission.status == 'ungraded':
+                        ungradedcount = ungradedcount + 1
+            for worksheet in self.course.worksheets.all():
+                if worksheet.last_submission(student):
+                    attemptcount = attemptcount + worksheet.last_submission(student).get_number()
+                worksheetcount = worksheetcount + 1
+
+            worksheetsdict[student.user.username] = worksheetcount
+            ungradedcountdict[student.user.username] = ungradedcount
+            attemptsdict[student.user.username] = attemptcount
```

Comment:
Check out django's annotate method https://docs.djangoproject.com/en/1.11/topics/db/aggregation/
I think the way that we get this info could be a lot more efficient if we used that. Basically in this loop we hit the database 2n times where n is the number of students in the current course. I think by annotating, we could get this down to 1 query, or at least constant complexity.

Reply:
Something like this :
```python
students = self.course.students.all()
students.annotate(num_submissions=Count("studentsubmission", filter=Q(studentsubmission__worksheet__course=self.course)))
students.annotate(num_ungraded_submissions=Count("studentsubmission", filter=Q(studentsubmission__worksheet__course=self.course) & Q(studentsubmission__status='ungraded'))) 
students.annotate(num_worksheets=Count("expression__worksheet", filter=Q(expression__worksheet__course=c)))
```